Pathway vectorstore and rag-pathway template

janchorowski opened the pull request:
- **Description:** Integration with pathway.com data processing pipeline acting as an always updated vectorstore
  - **Issue:** not applicable
  - **Dependencies:** optional dependency on [`pathway`](https://pypi.org/project/pathway/)
  - **Twitter handle:** pathway_com

The PR provides and integration with `pathway` to provide an easy to use always updated vector store:

```python
import pathway as pw
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.text_splitter import CharacterTextSplitter
from langchain.vectorstores import PathwayVectorClient, PathwayVectorServer

data_sources = []
data_sources.append(
    pw.io.gdrive.read(object_id="17H4YpBOAKQzEJ93xmC2z170l0bP2npMy", service_user_credentials_file="credentials.json", with_metadata=True))

text_splitter = CharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
embeddings_model = OpenAIEmbeddings(openai_api_key=os.environ["OPENAI_API_KEY"])
vector_server = PathwayVectorServer(
    *data_sources,
    embedder=embeddings_model,
    splitter=text_splitter,
)
vector_server.run_server(host="127.0.0.1", port="8765", threaded=True, with_cache=False)
client = PathwayVectorClient(
    host="127.0.0.1",
    port="8765",
)
query = "What is Pathway?"
docs = client.similarity_search(query)
```

The `PathwayVectorServer` builds a data processing pipeline which continusly scans documents in a given source connector (google drive, s3, ...) and builds a vector store.  The `PathwayVectorClient` implements LangChain's `VectorStore` interface and connects to the server to retrieve documents.

lewymati committed:
Pathway vectorstore and rag-pathway template

---------

Co-authored-by: mlewandowski <mlewandowski@MacBook-Pro-mlewandowski.local>
Co-authored-by: Berke <berkecanrizai1@gmail.com>
Co-authored-by: Jan Chorowski <janchorowski@users.noreply.github.com>
Co-authored-by: Adrian Kosowski <adrian@pathway.com>

hwchase17 commented on libs/community/langchain_community/vectorstores/pathway.py:
```diff
@@ -0,0 +1,141 @@
+
+from typing import Callable, List, Optional
+
+import pathway as pw
```
this should be a conditional import

janchorowski replied:
Sorry about the omission, all should be optional now.

hwchase17 reviewed (commented):
lets not import this into langchain - langchain should remain unchanged

only langchain-community should be updated, and we should import directly from there

lewymati committed:
fix imports (#6)

Change to a conditional import.

---------

Co-authored-by: mlewandowski <mlewandowski@macbook-pro-mlewandowski.home>

berkecanrizai committed:
Berke/fix docs (#7)

Fix documentation markdown formatting.

---------

Co-authored-by: mlewandowski <mlewandowski@macbook-pro-mlewandowski.home>

lewymati committed:
Merge branch 'langchain-ai:master' into master

Commit:
revert pyproject/poetry.lock

Commit:
optional

Commit:
update pyproject and lockfile

It was done as follows:
1. fetch fresh langchain master
2. `poetry add --optional pathway@latest --python ">=3.10"`
3. `poetry lock --no-update`

janchorowski commented:
@hwchase17 we have fixed poetry lock and used type annotations suitable for Py3.8, can you re-trigger the CI run?

lewymati committed:
Merge branch 'master' into master

Commit:
update poetry.lock hash

lewymati committed:
Merge branch 'master' into master

Commit:
merge

janchorowski committed:
Add newline required by Ruff.

janchorowski commented:
@efriis I tried to fix the formatting, now CI should be clean.

Commit:
fix docstring

hwchase17 committed:
cr

hwchase17 committed:
cr

lewymati committed:
scores for pathway vectorstore + self-query retrieval (#8)

* scores for pathway vectorstore

* .

* fix vectorstore client import

* add self-query retrieving using Pathway

* update template

* lint

* improve integration test

* update minimum pathway version (support for scores)

---------

Co-authored-by: mlewandowski <mlewandowski@macbook-pro-mlewandowski.home>

berkecanrizai committed:
feat: documentation notebook item ordering

berkecanrizai committed:
fix: typo

janchorowski committed:
Fix formatting.

hwchase17 commented on libs/langchain/langchain/vectorstores/pathway.py:
```diff
@@ -0,0 +1,6 @@
+from langchain_community.vectorstores.pathway import (
```
as mentioned, lets not import this into langchain - langchain should remain unchanged

janchorowski replied:
Sorry, we did it mimicking how the other vectorstores were integrated.

However, if we remove this import, how should we proceed with the self-query retriever? Similarly to other vectorstores, we have extended the mapping of stores to translators at https://github.com/langchain-ai/langchain/pull/14859/files#diff-d9f3d054efd12eda9d8d1022d9d7023eb8579c3f4842a051b5f459927f4f1426R74 but this also changes core LangChain. What do you suggest @hwchase17?

janchorowski replied:
@hwchase17 I've noticed that the other vectorstores are now imported from langchain_community - I've changed the PR accordingly.

janchorowski committed:
Remove imports from outside langchain_community.

janchorowski committed:
Merge remote-tracking branch 'upstream/master'

hwchase17 reviewed (approved):
looks reasonable to me. thanks!

janchorowski committed:
ruff formatting

janchorowski commented:
@hwchase17 Thanks for the review!
@efriis can you re-trigger CI? Locally I verified that the following run:
- `make lint`
- `make spellcheck`
- `make test` under both community and libs/langchain

janchorowski committed:
Merge remote-tracking branch 'upstream/master'

janchorowski committed:
Merge remote-tracking branch 'upstream/master'

janchorowski commented:
Dear @hwchase17 @efriis @baskaryan can you help me push it through the finish line? I've just merged in latest master and made sure linters and tests work locally.

janchorowski committed:
Merge remote-tracking branch 'upstream/master'

baskaryan committed:
Merge branch 'master' into janchorowski/master

baskaryan committed:
cr

Commit:
fix self query test to be immune to different timezones

lewymati commented:
We fixed the test which was working locally but failed at CI due to the different timezone.
Please accept our apologizes.

@baskaryan Could you please re-trigger the CI once again? We hope there won't be any other issues.
Thanks for your understanding.

efriis commented on libs/langchain/langchain/retrievers/self_query/base.py:
```diff
@@ -12,6 +12,7 @@
     MongoDBAtlasVectorSearch,
     MyScale,
     OpenSearchVectorSearch,
+    PathwayVectorClient,
```
this will break backwards compatibility for anyone importing from langchain as harrison mentioned above.

efriis reviewed (dismissed):
Let's remove the self query retriever and template from this PR and start with just the vector store.

Next, maybe try to add a generic implementation of self-query. This won't break backwards compatibility by importing something new from community into langchain.

And finally we can do a third PR with the new template.

janchorowski committed:
Merge remote-tracking branch 'upstream/master'

janchorowski committed:
Remove self-query

janchorowski commented:
@efriis I've removed the self-query. The template didn't use it (it only uses the vector_store) so I kept it. Can you re-review?

janchorowski committed:
Merge remote-tracking branch 'upstream/master'

janchorowski commented:
@efriis @baskaryan I've updated the PR. Please run CI.

efriis commented on libs/community/langchain_community/vectorstores/pathway.py:
```diff
@@ -0,0 +1,187 @@
+)
+
+
+class PathwayVectorServer:
```
Let's remove this and instead document how to run it separately in the docs

janchorowski replied:
@efriis where would you propose to move it?

Unlike a classical vector database, the pathway vector store server is a single process that runs a full document ingestion pipeline and index accessible using a REST API, so it replaces a database + processing pilpeline + orchestrator. The `PathwayVectorServer` class in `libs/community/langchain_community/vectorstores/pathway.py` is a wrapper that allows easy usage of Langchain's document processing classes. I think it is beneficial - LangChain users get an easy way to build always up-to-date chatbots without having to learn another set of document processing tools.

efriis replied:
Hey there! Let's just have the client in community, and you can document how to run the backend server in the docs + the template readme?

efriis commented on templates/rag-pathway/README.md:
add instructions on running th epathway server

efriis reviewed (commented):
1 removal and request for documentation on running server! Sorry for the delay on this one

janchorowski committed:
Merge remote-tracking branch 'upstream/master'

janchorowski committed:
Add README instructions on running the standalone server

janchorowski reviewed (commented):
@efriis thanks for the re-review!

I have added instructions on how to run the server to the template (by default it runs in a thread and `langchain serve` is the only command a user needs to run, but there is now an option to also run it in a separate process.

efriis reviewed (commented):
Hey @janchorowski ! Also repeating my previous request to remove the template from this PR, and do it as a separate PR.

Currently, the template is locked to a current published version of `langchain-community`, which does not have the pathway vector store. So the steps would be:

1. address comments
2. merge this pr
3. publish community
4. put up a new PR with the template, with a minimum langchain/community version that has the pathway vector store

Let me know if that doesn't make sense!

efriis commented:
Also responded up [here](https://github.com/langchain-ai/langchain/pull/14859#discussion_r1509335205). The backend running code should not be in the package.

szymondudycz committed:
Remove VectorStoreServer from integration (#10)

janchorowski committed:
Fix docstring

janchorowski committed:
Merge branch 'master' into master

janchorowski commented:
@efriis we have simplified the PR, leaving only the client and changing the instruction for a quick start using a publicly available server, then pointing to instructions on how to run it. Please review!

janchorowski committed:
Merge remote-tracking branch 'upstream/master'

janchorowski committed:
linters

janchorowski commented:
@efriis I fixed linters

szymondudycz committed:
Merge branch 'master' into update_imports

janchorowski commented:
@efriis please trigger CI, we resolved a merge conflict

janchorowski committed:
Merge remote-tracking branch 'upstream/master' into update_imports

janchorowski committed:
Ruff lint

janchorowski commented:
@efriis @baskaryan sorry to bother you, I re-merged master again and rerun linters. On my end locally `make lint` works, `make test` fails with `FAILED tests/unit_tests/callbacks/test_callback_manager.py::test_callback_manager_configure_context_vars - AttributeError: 'Client' object has no attribute 'tracing_queue'` which seems unrelated and hopefully won't block this more.

baskaryan committed:
Merge branch 'master' into update_imports
